### PR TITLE
Periodicity fix for global unstructured grids

### PR DESCRIPTION
**Pull request by sbrus89**
- Corrects calculation of element areas and edge lengths across the
   dateline in w3triamd.ftn

 - Fixes calculation to determine whether a point is inside an element
   that spans the dateline in w3triamd.ftn

 - Corrects calculation of element centers and corners across dateline
   in wmscrpmd.ftn

 - Write out SCRIP file in netCDF format at the end of ww3_grid
   (for creating offline remaping files). This required adding netCDF support
   to ww3_grid

 - Also includes minor fix for the /O7a switch in w3iopomd.ftn

### CODE REVIEW

**Comment by sbrus89** [comment]
Hi WAVEWATCHIII developers! This is my first PR to the repo, so please let me know what I can do to make the review process go smoothly. I appreciate all your feedback.

**Comment by sbrus89** [comment]
I am not sure if the SCRIP output functionality in `ww3_grid` is a feature that you all want at this point. I included it in the PR for the sake of discussion. Also, it may need to be put under a different switch. I'm open to suggestions here.

**Comment by sbrus89** [comment]
Also, `w3triamd.ftn` and `wmscrpmd.ftn` both have `FIX_PERIODICITY` subroutines that are essentially the same. I did this in order to avoid creating additional module dependencies, but am happy to address this based on your comments.

**Comment by aliabdolali** [comment]
Hi @sbrus89 
Thanks for sending this PR. I have read your GMD paper with @JessicaMeixner-NOAA 
BTW, I have a fix for periodicity for global unstructured grids on my personal fork but have not had a chance to do the final tests and merge it to the mainstream.
https://github.com/aliabdolali/WW3/tree/FB_Global
It is branched out from our development with USACE on the efficiency improvement of unstructured WW3 and implicit scheme implementation. 
Let me take a look at your development, it might be better than mine. I think we can merge it after a few iterations.

File: model/ftn/ww3_grid.ftn [hunk 1]
```diff
@@ -523,6 +523,18 @@
                           NOSWLL, PTMETH, PTFCUT
       USE W3TIMEMD, ONLY: CALTYPE
       USE W3NMLGRIDMD
+!/SCRIP      USE SCRIP_GRIDS, ONLY: GRID1_UNITS, GRID1_NAME, &
+!/SCRIP                             GRID1_CENTER_LON, GRID1_CENTER_LAT, &
+!/SCRIP                             GRID1_CORNER_LON, GRID1_CORNER_LAT, &
+!/SCRIP                             GRID1_MASK, GRID1_SIZE, GRID1_RANK, &
+!/SCRIP                             GRID1_IMASK, &
+!/SCRIP                             GRID1_CORNERS, GRID1_DIMS
+!!/SCRIP      USE SCRIP_REMAP_VARS
```
Comment (JessicaMeixner-NOAA) [inline model/ftn/ww3_grid.ftn:532]: Remove commented out lines here and below?

**Comment by aliabdolali** [comment]
Hi @sbrus89, 
We have a couple of active applications/developments for the global unstr case, so I decided to prepare a regtest for it. I'll push it to your PR in the next few days. One thing that I noticed is that although the defined points for the point outputs (in ww3_shel.inp) are inside the domain, they are skipped. See for example:
```
*** WAVEWATCH-III WARNING :
     OUTPUT POINT OUT OF GRID :    -86.417    45.329  45002
     POINT SKIPPPED
```
We need to fix it.
I'll keep you posted on further progress/checks.
AA

**Comment by sbrus89** [comment]
Hi @aliabdolali, 

Thanks for setting up the regtest and for catching the skipped points. I like aspects of your implementation better, so feel free to blend the two as needed. Let me know what I can do to help.

Thanks again,
Steven

**Comment by aliabdolali** [comment]
Hi @sbrus89 
I merged my development into your work. Basically, the main difference is that the global grids are not necessarily between [-180 180], so I made it more general, therefore the \delta between min/max of longitudes is 360.

I also added a regtests. I checked the periodicity and it is working fine (see regtests/ww3_tp2.21/info). Then I shifted the mesh from [-180 180 ] to [-360 0] and checked it again. it is working fine. Please try it at your end and let me know if you need changes. 

I pushed the large files in regtests/ww3_tp2.21/input but need to add them to our tar file on ftp. Once you are happy with it, I will merge the development to your PR, move the large files to ftp and will check the entire regtests. Hope all the tests pass, otherwise, I'll contact you.

Also, there is a comment from Jessica, could you remove those lines?

Please add some comments about scrip.nc. What is the purpose of it?

Regards
AA

**Comment by sbrus89** [comment]
Thanks for your help with this, @aliabdolali. I like your cleaner and more general approach. I am working testing this and will get back with you soon.

The `scrip.nc` file is an SCRIP input file that can be used with ESMF_regrid_gen_weights (or similar). I added this to aid in the creation of coupler mapping files for unstructured grids for E3SM. I wasn't sure if this was something that would be useful to include in the PR or not, so I decided to leave it in just in case. We can certainly leave it out if you think that would be better.

File: model/ftn/wmscrpmd.ftn [hunk 2]
```diff
@@ -697,12 +731,15 @@
...
+          ELON=(ELON1 + ELON2 + ELON3)/3.0
+          ELAT=(ELAT1 + ELAT2 + ELAT3)/3.0
+       
+          
           LON_CENT_TRIG(IE)=ELON
           LAT_CENT_TRIG(IE)=ELAT
+
         END DO
-        DODEBUG=0
+        DODEBUG=1
```
Comment (aliabdolali) [inline model/ftn/wmscrpmd.ftn:742]: @sbrus89 Can we turn it back to DODEBUG=0?

File: model/ftn/wmscrpmd.ftn [hunk 3]
```diff
@@ -1065,31 +1103,46 @@
      &   GRID_CORNER_LON, GRID_CORNER_LAT, GRID_MASK,                   &
      &   GRID_DIMS, GRID_SIZE, GRID_CORNERS, GRID_RANK)
       END IF
-      CHECKSIGNS=0
+      CHECKSIGNS=1
```
Comment (aliabdolali) [inline model/ftn/wmscrpmd.ftn:1106]: @sbrus89 Is it a part of your development/debug or it should be always checked? if not, please turn it back to 0 or let me know if you want to keep it.

File: model/ftn/wmscrpmd.ftn [hunk 4]
```diff
@@ -1065,31 +1103,46 @@
...
+            ELSE IF (THEDET.LT.-1d-8) THEN
               NBMINUS=NBMINUS+1
+            ELSE
+              NBZERO=NBZERO+1
             END IF
           END DO
         END DO
-        WRITE(*,*) 'SI nbPlus=', nbPlus, ' nbMinus=', nbMinus
+
+        PRINT*, 'SI nbPlus=', nbPlus, ' nbMinus=', nbMinus, ' nbZero=', nbZero
```
Comment (aliabdolali) [inline model/ftn/wmscrpmd.ftn:1144]: @sbrus89  What about this one? is it for debugging?

File: model/ftn/ww3_grid.ftn (hunk 1 above)
Comment (aliabdolali) [reply model/ftn/ww3_grid.ftn:532]: @sbrus89 could you take care of Jessica's comments about outlines?

File: model/ftn/ww3_grid.ftn [hunk 5]
```diff
@@ -523,6 +523,18 @@
...
+!!/SCRIP      USE SCRIP_CONSTANTS
+!/SCRIP      USE SCRIP_KINDSMOD
+!!/SCRIP      USE SCRIP_INTERFACE
```
Comment (aliabdolali) [inline model/ftn/ww3_grid.ftn:535]: @sbrus89 here and elsewhere there are outlines.

**Pull request closed without merging**